src/legacy-routes.js

```diff
@@ -14,36 +19,108 @@ const invalidTx = 'Invalid transaction id!'
+const wrapHashPrefix = (hash: string): string => `\\x${hash}`
+const unwrapHashPrefix = (hash: string): string => hash.substr(2)
 
-const combinedBalance = (transactions, addresses) => {
-  const addressSet = new Set(addresses)
-  const totalIn = transactions.reduce((acc, tx) =>
-    acc.plus(txAddressCoins(tx.outputs_address, tx.outputs_amount, addressSet)), Big(0))
-  const totalOut = transactions.reduce((acc, tx) =>
-    acc.plus(txAddressCoins(tx.inputs_address, tx.inputs_amount, addressSet)), Big(0))
-  return totalIn.sub(totalOut)
+/**
+ * Initializes tx entry in the caTxList format into which inputs can be added and summed.
+ * @param {Tx} tx Transaction from database
+ */
+const initializeTxEntry = (tx: Tx) : TxEntry => ({
+  ctbId: unwrapHashPrefix(tx.hash),
+  ctbTimeIssued: moment(tx.time).unix(),
+  ctbInputs: [],
+  ctbOutputs: [],
+  ctbInputSum: { getCoin: Big(0) },
+  ctbOutputSum: { getCoin: Big(0) },
+})
+
+/**
+ * Initializes tx input/output entry
+ * @param {Movement} movement Tx input or output
+ */
+const initializeMovementEntry = (movement: Movement)
+: MovementEntry => [movement.address, { getCoin: movement.value }]
+
+/**
+ * Pushes tx movement to corresponding tx entry and sums its value
+ * according to whether it's tx input or output
+ * @param {Map<number, TxEntry>} txMap Map which holds txEntries at id keys used for fast lookups
+ * @param {Movement} movement Tx input or output
+ * @param {boolean} isInput Indicates whether to add movement to tx inputs or outputs
+ */
+const pushMovementToTxMap = (
+  txMap: Map<number, TxEntry>, movement: Movement, isInput: boolean,
+) : void => {
+  const txEntry = txMap.get(movement.txid)
+  if (!txEntry) { return }
+
+  if (isInput) {
+    txEntry.ctbInputs.push(initializeMovementEntry(movement))
+    txEntry.ctbInputSum.getCoin = txEntry.ctbInputSum.getCoin.plus(movement.value)
+  } else {
+    txEntry.ctbOutputs.push(initializeMovementEntry(movement))
+    txEntry.ctbOutputSum.getCoin = txEntry.ctbOutputSum.getCoin.plus(movement.value)
+  }
 }
 
-const txToAddressInfo = (row) => ({
-  ctbId: row.hash,
-  ctbTimeIssued: moment(row.time).unix(),
-  ctbInputs: row.inputs_address.map(
-    (addr, i) => [addr, { getCoin: row.inputs_amount[i] }]),
-  ctbOutputs: row.outputs_address.map(
-    (addr, i) => [addr, { getCoin: row.outputs_amount[i] }]),
-  ctbInputSum: {
-    getCoin: `${arraySum(row.inputs_amount)}`,
-  },
-  ctbOutputSum: {
-    getCoin: `${arraySum(row.outputs_amount)}`,
-  },
-})
+/**
+ * Assigns tx inputs and outputs to corresponding transactions to build caTxList
+ * @param {Array<Tx>} transactions Array of transactions fetched from database
+ * @param {Array<Movement>} txInputs Array of tx inputs
+ * @param {Array<Movement>} txOutputs Array of tx outputs
+ */
+const buildTxList = (
+  transactions: Array<Tx>, txInputs: Array<Movement>, txOutputs: Array<Movement>,
+) => {
+  const txMap = new Map(transactions.map(tx => [tx.id, initializeTxEntry(tx)]))
```

Comment:
I think it would be cleaner to have the `initializeTxEntry` take also the array of txInputs and txOutputs. And you can leverage lodash groupBy function to get a map where the inputs and outputs woud be in an object keyed by txId: https://lodash.com/docs/2.4.2#groupBy